mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp

```diff
@@ -423,148 +321,138 @@ void CHudSpeedMeter::OnThink()
-
-                // Update the compare color to have the alpha defined by animations (fade out if 5+ sec)
-                compareColorFade.SetColor(compareColor.r(), compareColor.g(), compareColor.b(), m_fStageStartAlpha);
-
+                m_pStageEnterExitComparisonLabel->SetPos(0, m_pStageEnterExitLabel->GetYPos());
+                m_pStageEnterExitComparisonLabel->SetFgColor(Color(compareColor.r(), compareColor.g(), compareColor.b(), m_fStageStartAlpha));
+                m_pStageEnterExitComparisonLabel->SetText(enterVelComparisonStr);
                 // Add the compare string to the length
-                combinedLength += spaceBetweenLabels + UTIL_ComputeStringWidth(labelFont, enterVelComparisonRounded);
+                combinedLength += UTIL_ComputeStringWidth(labelFont, enterVelComparisonStr);
+            }
+            else
+            {
+                // ignore ending space as comparison is not being shown
+                combinedLength -= UTIL_ComputeStringWidth(labelFont, " ");
             }
-            int offsetXPos = 0 - ((GetWide() - combinedLength) / 2);
 
-            // split_xpos = GetWide() / 2 - (enterVelANSIWidth + increaseX) / 2;
+            int offsetXPos = 0 - ((GetWide() - combinedLength) / 2);
             m_pStageEnterExitLabel->SetPos(offsetXPos, m_pStageEnterExitLabel->GetYPos());
-            m_pStageEnterExitLabel->SetFgColor(actualColorFade);
-            m_pStageEnterExitLabel->SetText(enterVelRounded);
+            m_pStageEnterExitLabel->SetFgColor(Color(enterColor.r(), enterColor.g(), enterColor.b(), m_fStageStartAlpha));
+            m_pStageEnterExitLabel->SetText(enterVelStr);
+        }
+    }
+}
 
-            // print comparison as well
-            if (loadedComparison)
-            {
-                m_pStageEnterExitComparisonLabel->SetPos(spaceBetweenLabels, m_pStageEnterExitLabel->GetYPos());
-                m_pStageEnterExitComparisonLabel->SetFgColor(compareColorFade);
-                m_pStageEnterExitComparisonLabel->SetText(enterVelComparisonRounded);
-            }
-            else // only print velocity
-            {
-                m_pStageEnterExitComparisonLabel->SetText("");
-            }
+void CHudSpeedMeter::SetLabelHeight(bool isActive, Label *label, int height)
+{
+    label->SetAutoTall(isActive);
+    label->SetTall(isActive ? height : 0);
+    if (!isActive) label->SetText("");
+}
+
+void CHudSpeedMeter::SetSpeedometerLabel(bool isActive, Label *label, int height, int roundedSpeed, Color color, float alpha)
+{
+    if (isActive)
+    {
+        char speedoValue[BUFSIZELOCL];
+        Q_snprintf(speedoValue, sizeof(speedoValue), "%i", roundedSpeed);
+        label->SetText(speedoValue);
+        if (alpha >= 0.0f) // apply fade if specified
+        {
+            color = Color(color.r(), color.g(), color.b(), alpha);
+        }
+        label->SetFgColor(color);
+    }
+    SetLabelHeight(isActive, label, height);
+}
```

Comment:
This is really smelling like a custom class deriving from Label is needed...

Something like `SpeedometerLabel` that has all of these parameters assigned as member variables (could even trade isActive for a `ConVar *` to the visibility toggle and query it in this method in the new class).

I say this because the name "SetSpeedometerLabel" is very... *underselling* what it is actually doing, and all of these values can be set beforehand and not strung along, then chained into this master setter method.

Reply:
You could even move the colorization there, getting rid of the need to "do everything twice" as I've seen throughout the OnThink method.

Reply:
And you could even do calculations like `CalculateFromVelocity(float velInput);` and do most of the logic from OnThink in there, stuff like conversion of units, etc. You'd have control over feeding it from the parent speedometer panel (this class file).

I'm thinking a base `SpeedometerLabel` class and then maybe overriding as need be, since there seems to be different label requirements (abs and horizontal [and eventually ramp speeds] constantly show, jump/enter show then fade, etc). Could have the base class have "paint behaviors" that get set at construction ("show constantly" / "show then fade out") and you won't need to make overrides.

I hope this is making some sense. This refactor is going in the right direction, and definitely opening my eyes on possibilities to further clean it up and make it expandable for the other speeds needed in the future.

Reply:
Having now tried this, I couldn't get an implementation I was happy with. The calculation of the actual value, it's colorization, whether to adjust for replay timescales (last jump & stage enter don't need to be adjusted), unit conversions (only abs/horiz use energy, and they end up being the same value in that case anyways).

I ended up trying to use callback functions for the velocity calculation & colorization in the `SpeedometerLabel` constructor but it became apparent that every speedometer had their own behavior. Only the abs/horiz speedometers are similar.

Stage enter velocity also didn't fit in well as it comes across as a string not a float/integer. Maybe requires some other function in the comparison code to account for that, though I'm not sure what exactly the scope of this should be.